### PR TITLE
Scatter packet

### PR DESCRIPTION
**Pull request by ffoucart**
## Proposed changes

Add function to scatter packets at constant fluid-frame energy

### Upgrade instructions

NA

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by ffoucart** [comment]
Built on top of #5820

File: src/Evolution/Particles/MonteCarlo/EvolveMonteCarloPackets.cpp [hunk 1]
```diff
@@ -117,7 +115,7 @@ void diffuse_packet(gsl::not_null<Packet*> /*packet*/,
 
 void evolve_packets(
     gsl::not_null<std::vector<Packet>*> packets, const double& final_time,
-    const Mesh<3>& mesh,
+    std::mt19937& random_number_generator, const Mesh<3>& mesh,
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/EvolveMonteCarloPackets.cpp]: can you make the `random_number_generator` a `gsl::not_null<std::mt19937*>`?

File: src/Evolution/Particles/MonteCarlo/ScatteringMonteCarlo.hpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+struct Fluid;
+} // namespace Frame
+
+/// Items related to the evolution of particles
+/// Items related to Monte-Carlo radiation transport
+namespace Particles::MonteCarlo {
+
+struct Packet;
+
+// Elastic scattering, which is just redrawing the momentum from
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/ScatteringMonteCarlo.hpp]: `///` or `/*!` for doxygen comment

File: src/Evolution/Particles/MonteCarlo/ScatteringMonteCarlo.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ScatteringMonteCarlo.hpp"
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/ScatteringMonteCarlo.cpp]: `#include "Evolution/Particles/MonteCarlo/ScatteringMonteCarlo.hpp"`

File: src/Evolution/Particles/MonteCarlo/ScatteringMonteCarlo.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+#include "Evolution/Particles/MonteCarlo/MonteCarloPacket.hpp"
+
+namespace Particles::MonteCarlo {
+
+void scatter_packet(
+    gsl::not_null<Packet*> packet, std::mt19937& random_number_generator,
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/ScatteringMonteCarlo.cpp]: `const gsl`, make `std::mt...` a `const gsl::not_null<std::mt...*>`

File: src/Evolution/Particles/MonteCarlo/ScatteringMonteCarlo.cpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    const double& fluid_frame_energy,
+    const Jacobian<DataVector, 4, Frame::Inertial, Frame::Fluid>&
+        inertial_to_fluid_jacobian,
+    const InverseJacobian<DataVector, 4, Frame::Inertial, Frame::Fluid>&
+        inertial_to_fluid_inverse_jacobian) {
+  const size_t& idx = packet->index_of_closest_grid_point;
+  std::array<double, 4> four_momentum_fluid{fluid_frame_energy, 0.0, 0.0, 0.0};
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/ScatteringMonteCarlo.cpp]: I would move this down to line 28 and do:
```cpp
const std::array<double, 4> four_momentum_fluid{fluid_frame_energy,
sin_theta * cos(phi) * fluid_frame_energy,
sin_theta * sin(phi) * fluid_frame_energy,
cos_theta * fluid_frame_energy};
```
(pardon my poor indentation)

File: src/Evolution/Particles/MonteCarlo/ScatteringMonteCarlo.cpp [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  packet->momentum_upper_t = 0.0;
+  for (size_t d = 0; d < 4; d++) {
+    packet->momentum_upper_t +=
+        inertial_to_fluid_inverse_jacobian.get(0, d)[idx] *
+        gsl::at(four_momentum_fluid, d);
+  }
+  for (size_t d = 0; d < 3; d++) {
+    // Multiply by -1.0 because p_t = -p^t in an orthonormal frame
+    packet->momentum.get(d) = (-1.0) * gsl::at(four_momentum_fluid, 0) *
```
Comment (nilsdeppe) [inline src/Evolution/Particles/MonteCarlo/ScatteringMonteCarlo.cpp]: I think `-gsl::at(four_momentum_fluid, 0)` should compile fine without the multiplication

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Comment by ffoucart** [comment]
Last commit did NOT address code review comments; only rebased against latest branch (I pushed the wrong branch...)

**Comment by ffoucart** [comment]
Ok, now the last commit is meant to address code review...

**Review by kidder — COMMENTED** [review]
you can squash into a single commit

**Comment by ffoucart** [comment]
Large number of seg-faulting tests in clang 14 DEBUG... is this a known issue?

**Comment by wthrowe** [comment]
> Large number of seg-faulting tests in clang 14 DEBUG... is this a known issue?

Yes.  Things are going wrong with that configuration on a number of PRs.  Don't know why yet.

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**